feat(cli): allow runner configuration to be an object with cwd and args

mhbagheri-99 opened the pull request:
This PR implements the feature requested in #13779 to allow the "runner" configuration in `tauri.conf.json` to be either a string (existing behavior) or an object with additional properties.

## Changes

- Enhanced the runner configuration to support an object format with the following properties:
  - `cmd`: The command to execute (equivalent to the previous string format)
  - `cwd`: Optional working directory to run the command from
  - `args`: Optional array of arguments to pass to the runner

## Example Usage

The runner can now be configured as an object:

```json
{
  "build": {
    "runner": {
      "cmd": "my_custom_cargo",
      "cwd": "/path/to/project",
      "args": ["--quiet", "--offline"]
    }
  }
}
```

Or continue using the existing string format for backwards compatibility:

```json
{
  "build": {
    "runner": "cargo"
  }
}
```

## Backwards Compatibility

This change maintains full backwards compatibility - existing configurations using a string for the runner will continue to work as before.

Closes #13779

---

mhbagheri-99 committed:
Update config.schema.json

mhbagheri-99 committed:
Add RunnerConfig for customizable build runner

Replaces runner String with RunnerConfig in CLI and config, allowing advanced runner configuration via string or object with cmd, cwd, and args. Updates schema and usage to support new format, and adds tests for serialization, deserialization, and API. Enables more flexible build and run command customization.

mhbagheri-99 committed:
Create runner-object-config.md

mhbagheri-99 committed:
Remove unused RunnerConfig import in tests

Cleaned up the test module in config.rs by removing the unused RunnerConfig import from two test functions.

mhbagheri-99 committed:
Merge branch 'dev' into improve-runner-configuration

mhbagheri-99 committed:
Fix tests failing

Updates related tests in tauri-utils to improve readability and maintain consistency. Minor import reordering in tauri-cli for clarity.

mhbagheri-99 committed:
Merge branch 'dev' into improve-runner-configuration

Legend-Master commented on crates/tauri-utils/src/config.rs:
```diff
@@ -3824,4 +3847,283 @@ mod test {
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+#[cfg_attr(feature = "schema", derive(JsonSchema))]
+#[serde(untagged)]
+pub enum RunnerConfig {
```
Should we move these to above, like close to `HookCommand`?

mhbagheri-99 replied:
@Legend-Master moved the implementation to make more sense according to your comment 🙂

Legend-Master reviewed (dismissed):
Thanks! Just a nitpick

mhbagheri-99 committed:
Move RunnerConfig enum and impls above BuildConfig

Relocated the RunnerConfig enum and its associated implementations to appear before the BuildConfig definition. This improves code organization and logical grouping of configuration-related types.

mhbagheri-99 committed:
Merge branch 'dev' into improve-runner-configuration
